File: api/loadtest/ingest.js
// M5.4 ingest profile: 50 POSTs/s of 100-result envelopes, for the mixed run.
//
//   k6 run loadtest/ingest.js
//
// Each iteration invents a fresh run_id, so every POST is a real insert of 100 rows rather
// than a duplicate that the row-level conflict rule would skip. At 50/s for 60 s that is
// 300,000 rows, which is more than the read profile's dataset - the run is short and the
// table is expected to grow.
import http from 'k6/http';
import { check } from 'k6';
import { BASE, discover } from './lib.js';

const RATE = Number(__ENV.INGEST_RATE || 50);
const DURATION = __ENV.DURATION || '60s';

export const options = {
    scenarios: {
        ingest: {
            executor: 'constant-arrival-rate',
            rate: RATE,
            timeUnit: '1s',
            duration: DURATION,
            preAllocatedVUs: 20,
            maxVUs: 100,
            tags: { scenario: 'ingest' },
        },
    },
    thresholds: {
        'http_req_duration{scenario:ingest}': ['p(95)<500'],
        'http_req_failed{scenario:ingest}': ['rate<0.001'],
        dropped_iterations: ['count==0'],
    },
};

export function setup() {
    const data = discover();
    // The machine block has to come back exactly as the schema wants it, and the machines
    // endpoint does not carry engine_version (that belongs to a run), so it is filled in.
    const res = http.get(`${BASE}/v1/machines`);
    const m = res.json('machines')[0];
    const machine = {
        id: m.id, hostname: m.hostname, cpu_model: m.cpu_model,
        physical_cores: m.physical_cores, logical_cpus: m.logical_cpus,
        l1d_kb: m.l1d_kb, l2_kb: m.l2_kb, l3_kb: m.l3_kb, memory_bytes: m.memory_bytes,
        os: m.os, kernel: m.kernel, compiler: m.compiler, compiler_flags: m.compiler_flags,
        engine_version: 'loadtest', engine_git_sha: 'loadtest',
    };
    if (m.virtualized) machine.virtualized = m.virtualized;
    return { machine, combos: data.combos };
}

function hex(n) {
    let s = '';
    for (let i = 0; i < n; i++) s += '0123456789abcdef'[Math.floor(Math.random() * 16)];
    return s;
}

// uuid4 builds the v4-shaped id the schema's pattern requires.
function uuid4() {
    return `${hex(8)}-${hex(4)}-4${hex(3)}-${'89ab'[Math.floor(Math.random() * 4)]}${hex(3)}-${hex(12)}`;
}

export default function (data) {
    const now = new Date();
    const stamp = (d) => d.toISOString().replace('Z', '000Z');
    const results = [];
    for (let i = 0; i < 100; i++) {
        results.push({
            workload: 'cpu_int', thread_count: 1, working_set_bytes: 0,
            metric: 'cpu_int_ops', value: 4.6e9 + Math.random() * 1e8, unit: 'ops/s',
            trial: i, timestamp: stamp(new Date(now.getTime() + i)), duration_ns: 50000000,
            params: { cold: 'clflush', pinned: true, warmup_trials: 5, trial_ms: 50 },
        });
    }
    const body = JSON.stringify({
        schema_version: 1, run_id: uuid4(),
        started_at: stamp(now), finished_at: stamp(new Date(now.getTime() + 5000)),
        machine: data.machine, argv: ['k6', 'loadtest/ingest.js'], results,
    });
    const res = http.post(`${BASE}/v1/runs`, body, {
        headers: { 'Content-Type': 'application/json' },
        tags: { name: 'ingest' },
    });
    check(res, {
        'status is 200': (r) => r.status === 200,
        'inserted 100': (r) => r.status === 200 && r.json('inserted') === 100,
    });
}

File: api/internal/store/ingest.go
package store

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"time"

	"github.com/jackc/pgx/v5"
	"github.com/jackc/pgx/v5/pgconn"

	"github.com/matthewlee/benchmark-platform/api/internal/model"
)

// copyBatchRows is how many measurement rows go into one CopyFrom call. It bounds the
// memory an ingest holds: results stream out of the JSON decoder into this batch and on
// to Postgres, so a 50K-result envelope is five flushes rather than one 90 MB slice.
const copyBatchRows = 10_000

// IngestResult is what POST /v1/runs answers with.
type IngestResult struct {
	RunID     string `json:"run_id"`
	Results   int    `json:"results"`
	Inserted  int    `json:"inserted"`
	Skipped   int    `json:"skipped"`
	Duplicate bool   `json:"duplicate"`
	MachineID string `json:"machine_id"`
}

var measurementCols = []string{
	"run_id", "machine_id", "workload", "metric", "thread_count", "working_set_bytes",
	"trial", "value", "unit", "recorded_at", "duration_ns", "params",
}

// IngestStream decodes the envelope from body and stores it, holding at most one batch of
// results in memory whatever the envelope's size. Chunks of one run (several POSTs sharing
// a run_id) append; a repeated chunk inserts nothing.
func (s *Store) IngestStream(ctx context.Context, body io.Reader, maxResults int) (IngestResult, error) {
	var ing *ingestion
	_, err := model.StreamRunEnvelope(body, maxResults,
		func(env *model.RunEnvelope, _ int, r *model.Result) error {
			if ing == nil {
				// The first result is where the run becomes real: the header is complete,
				// so the machine and run rows go in and the batches can start flowing.
				var err error
				if ing, err = s.beginIngest(ctx, env); err != nil {
					return err
				}
			}
			return ing.add(ctx, r)
		})
	if err != nil {
		if ing != nil {
			ing.rollback(ctx)
		}
		return IngestResult{}, err
	}
	if ing == nil {
		// StreamRunEnvelope rejects an empty results array, so this cannot normally happen.
		return IngestResult{}, errors.New("run envelope carried no results")
	}
	return ing.commit(ctx)
}

// IngestRun stores an already-decoded envelope. Used by the seeder and the tests; the
// HTTP handler streams instead.
func (s *Store) IngestRun(ctx context.Context, env *model.RunEnvelope) (IngestResult, error) {
	if err := env.Validate(); err != nil {
		return IngestResult{}, err
	}
	if len(env.Results) == 0 {
		return IngestResult{}, errors.New("run envelope carried no results")
	}
	ing, err := s.beginIngest(ctx, env)
	if err != nil {
		return IngestResult{}, err
	}
	for i := range env.Results {
		if err := ing.add(ctx, &env.Results[i]); err != nil {
			ing.rollback(ctx)
			return IngestResult{}, err
		}
	}
	return ing.commit(ctx)
}

// ingestion is one in-flight POST: a transaction, the run it belongs to, and the batch
// being filled.
type ingestion struct {
	tx        pgx.Tx
	runID     string
	machineID string
	rows      [][]any
	inserted  int
	skipped   int
	results   int
	// dedup routes rows through a temp table and ON CONFLICT DO NOTHING instead of copying
	// straight into `measurements`. It is on from the start when the run already existed
	// (a chunk or a retry) and switches on if a plain copy ever hits the unique index.
	dedup     bool
	tempReady bool
}

func (s *Store) beginIngest(ctx context.Context, env *model.RunEnvelope) (*ingestion, error) {
	started, err := model.ParseTimestamp(env.StartedAt)
	if err != nil {
		return nil, err
	}
	finished, err := model.ParseTimestamp(env.FinishedAt)
	if err != nil {
		return nil, err
	}
	argv, err := json.Marshal(env.Argv)
	if err != nil {
		return nil, fmt.Errorf("marshal argv: %w", err)
	}
	tx, err := s.pool.Begin(ctx)
	if err != nil {
		return nil, fmt.Errorf("begin: %w", err)
	}
	in := &ingestion{tx: tx, runID: env.RunID, machineID: env.Machine.ID,
		rows: make([][]any, 0, copyBatchRows)}

	if err := upsertMachine(ctx, tx, &env.Machine, started, finished); err != nil {
		in.rollback(ctx)
		return nil, err
	}
	tag, err := tx.Exec(ctx, `
		INSERT INTO runs (id, machine_id, engine_version, engine_git_sha,
		                  started_at, finished_at, argv)
		VALUES ($1, $2, $3, $4, $5, $6, $7)
		ON CONFLICT (id) DO NOTHING`,
		env.RunID, env.Machine.ID, env.Machine.EngineVersion, env.Machine.EngineGitSHA,
		started, finished, argv)
	if err != nil {
		in.rollback(ctx)
		return nil, fmt.Errorf("insert run: %w", err)
	}
	// The run already exists: this POST is either the next chunk of it or a retry of a
	// chunk already stored. Both are handled by the same row-level conflict rule.
	in.dedup = tag.RowsAffected() == 0
	return in, nil
}

func (in *ingestion) add(ctx context.Context, r *model.Result) error {
	ts, err := model.ParseTimestamp(r.Timestamp)
	if err != nil {
		return err
	}
	var params any
	if len(r.Params) > 0 {
		params = []byte(r.Params)
	}
	in.rows = append(in.rows, []any{
		in.runID, in.machineID, r.Workload, r.Metric, r.ThreadCount, r.WorkingSetBytes,
		r.Trial, r.Value, r.Unit, ts, r.DurationNs, params,
	})
	in.results++
	if len(in.rows) >= copyBatchRows {
		return in.flush(ctx)
	}
	return nil
}

func (in *ingestion) flush(ctx context.Context) error {
	if len(in.rows) == 0 {
		return nil
	}
	defer func() { in.rows = in.rows[:0] }()
	if in.dedup {
		return in.flushDedup(ctx)
	}
	// Fast path: this run is new, so nothing in the table can conflict with it - unless
	// the envelope repeats a configuration itself. A savepoint makes that recoverable
	// instead of poisoning the transaction.
	if _, err := in.tx.Exec(ctx, "SAVEPOINT batch"); err != nil {
		return fmt.Errorf("savepoint: %w", err)
	}
	n, err := in.tx.CopyFrom(ctx, pgx.Identifier{"measurements"}, measurementCols,
		pgx.CopyFromRows(in.rows))
	if err != nil {
		var pgErr *pgconn.PgError
		if !errors.As(err, &pgErr) || pgErr.Code != "23505" { // unique_violation
			return fmt.Errorf("copy measurements: %w", err)
		}
		if _, rbErr := in.tx.Exec(ctx, "ROLLBACK TO SAVEPOINT batch"); rbErr != nil {
			return fmt.Errorf("rollback to savepoint: %w", rbErr)
		}
		in.dedup = true // the envelope repeats a configuration; conflict-check from here on
		return in.flushDedup(ctx)
	}
	if _, err := in.tx.Exec(ctx, "RELEASE SAVEPOINT batch"); err != nil {
		return fmt.Errorf("release savepoint: %w", err)
	}
	in.inserted += int(n)
	return nil
}

// flushDedup copies the batch into a temp table and moves it across with ON CONFLICT DO
// NOTHING, because COPY itself cannot express a conflict rule.
func (in *ingestion) flushDedup(ctx context.Context) error {
	if !in.tempReady {
		if _, err := in.tx.Exec(ctx, `
			CREATE TEMP TABLE ingest_batch ON COMMIT DROP AS
			SELECT run_id, machine_id, workload, metric, thread_count, working_set_bytes,
			       trial, value, unit, recorded_at, duration_ns, params
			FROM measurements WITH NO DATA`); err != nil {
			return fmt.Errorf("create temp table: %w", err)
		}
		in.tempReady = true
	}
	if _, err := in.tx.Exec(ctx, `TRUNCATE ingest_batch`); err != nil {
		return fmt.Errorf("truncate temp table: %w", err)
	}
	if _, err := in.tx.CopyFrom(ctx, pgx.Identifier{"ingest_batch"}, measurementCols,
		pgx.CopyFromRows(in.rows)); err != nil {
		return fmt.Errorf("copy into temp table: %w", err)
	}
	tag, err := in.tx.Exec(ctx, `
		INSERT INTO measurements (run_id, machine_id, workload, metric, thread_count,
		                          working_set_bytes, trial, value, unit, recorded_at,
		                          duration_ns, params)
		SELECT run_id, machine_id, workload, metric, thread_count, working_set_bytes,
		       trial, value, unit, recorded_at, duration_ns, params
		FROM ingest_batch
		ON CONFLICT (run_id, metric, thread_count, working_set_bytes, trial) DO NOTHING`)
	if err != nil {
		return fmt.Errorf("insert from temp table: %w", err)
	}
	in.inserted += int(tag.RowsAffected())
	in.skipped += len(in.rows) - int(tag.RowsAffected())
	return nil
}

func (in *ingestion) commit(ctx context.Context) (IngestResult, error) {
	res := IngestResult{RunID: in.runID, MachineID: in.machineID}
	if err := in.flush(ctx); err != nil {
		in.rollback(ctx)
		return res, err
	}
	if err := in.tx.Commit(ctx); err != nil {
		return res, fmt.Errorf("commit: %w", err)
	}
	res.Results = in.results
	res.Inserted = in.inserted
	res.Skipped = in.skipped
	res.Duplicate = in.inserted == 0 && in.results > 0
	return res, nil
}

func (in *ingestion) rollback(ctx context.Context) {
	if in.tx != nil {
		_ = in.tx.Rollback(ctx)
	}
}

func upsertMachine(ctx context.Context, tx pgx.Tx, m *model.Machine, started, finished time.Time) error {
	_, err := tx.Exec(ctx, `
		INSERT INTO machines (id, hostname, cpu_model, physical_cores, logical_cpus,
		                      l1d_kb, l2_kb, l3_kb, memory_bytes, os, kernel, compiler,
		                      compiler_flags, virtualized, first_seen, last_seen)
		VALUES ($1,$2,$3,$4,$5,$6,$7,$8,$9,$10,$11,$12,$13,$14,$15,$16)
		ON CONFLICT (id) DO UPDATE SET
			hostname = EXCLUDED.hostname,
			cpu_model = EXCLUDED.cpu_model,
			physical_cores = EXCLUDED.physical_cores,
			logical_cpus = EXCLUDED.logical_cpus,
			l1d_kb = EXCLUDED.l1d_kb,
			l2_kb = EXCLUDED.l2_kb,
			l3_kb = EXCLUDED.l3_kb,
			memory_bytes = EXCLUDED.memory_bytes,
			os = EXCLUDED.os,
			kernel = EXCLUDED.kernel,
			compiler = EXCLUDED.compiler,
			compiler_flags = EXCLUDED.compiler_flags,
			virtualized = EXCLUDED.virtualized,
			first_seen = LEAST(machines.first_seen, EXCLUDED.first_seen),
			last_seen = GREATEST(machines.last_seen, EXCLUDED.last_seen)`,
		m.ID, m.Hostname, m.CPUModel, m.PhysicalCores, m.LogicalCPUs, m.L1dKB, m.L2KB,
		m.L3KB, m.MemoryBytes, m.OS, m.Kernel, m.Compiler, m.CompilerFlags, m.Virtualized,
		started, finished)
	if err != nil {
		return fmt.Errorf("upsert machine: %w", err)
	}
	return nil
}
